vignettes/fims-path-maturity.Rmd

```diff
@@ -0,0 +1,614 @@
+
+To evaluate maturity within the `Population` class, we first need to set up a shared pointer, linking the maturity module to population.
+
+```{Rcpp,  class.source = "fimschunk", eval = FALSE}
+//file: inst/include/population_dynamics/population/population.hpp
+// maturity is a shared pointer to MaturityBase
+int maturity_id = -999; /*!< id of maturity model object*/
+std::shared_ptr<fims::MaturityBase<Type>>
+      maturity; /*!< shared pointer to maturity module */
+```
+
+The `maturity_id` was set from R using the `SetMaturity()` method from the `PopulationInterface` class:
+```{r,  class.source = "fimschunk", eval = FALSE}
+fims$population$SetMaturity(maturity$get_id())
+```
+
+Notice above that population by default declares a pointer of type `MaturityBase` (parent class), not the specific maturity class we are using in this example, `LogisticMaturity` (child class). `MaturityBase` has an `evaluate` method with input arguments that match the inputs of each child class:
+
+```{Rcpp,  class.source = "fimschunk", eval = FALSE}
+//file: inst/include/population_dynamics/maturity/functors/maturity_base.hpp
+namespace fims {
+
+/** @brief Base class for all maturity functors.
+ *
+ * @tparam T The type of the maturity functor.
+ */
+
+template <typename T>
+struct MaturityBase : public FIMSObject<T> {
+  // id_g is the ID of the instance of the MaturityBase class.
+  // this is like a memory tracker.
+  // Assigning each one its own ID is a way to keep track of
+  // all the instances of the MaturityBase class.
+  static uint32_t id_g; /*!< The ID of the instance of the MaturityBase class */
+
+  /** @brief Constructor.
+   */
+  MaturityBase() {
+    // increment id of the singleton maturity class
+    this->id = MaturityBase::id_g++;
+  }
+
+  /**
+   * @brief Calculates the maturity.
+   * @param x The independent variable in the maturity function (e.g., logistic
+   * maturity at age or size).
+   */
+  virtual const T evaluate(const T& x) = 0;
+};
+```
+
+The evaluate function is a [virtual C++ function](https://www.geeksforgeeks.org/virtual-function-cpp/), which means the function can be overwritten by functions of the same name from a child class. This structure keeps the maturity module in population generic. We don't need any conditional statements to loop over all possible maturity formulations within population. We only need to add a new child maturity class with an Rcpp interface and we can automatically call it from population. There is a trade-off here. We are creating a nested hierarchical structure that makes the code base harder to read. In exchange, we're creating code with a lower [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity), that is there will only ever be one independent path from the user to the maturity->evaluate() call in population, regardless of how many maturity functions are added to FIMS. Code with lower cyclomatic complexity is easier to test, maintain, and extend.  
```

Comment:
@nathanvaughan-NOAA, @k-doering-NOAA, @Bai-Li-NOAA, does this paragraph capture what we discussed about virtual functions?

Reply:
This looks great! I just checked out the link and found the example and text there really effective. Thanks, Andrea!